### PR TITLE
pkg/trace: use msgp.Marshaler instead of msgp.Decodable of msgpack deserialization

### PR DESCRIPTION
**Pull request by bmermet**
### What does this PR do?

This PR changes the msgpack decoding implementation to use the `msgp.Marshaler` interface rather than `msgp.Decodable`. That's a tradeoff though, since we need to allocate memory for all the bytes of the received payload at once it will increase memory usage.

An extra benefit is that the utf8 normalization will no longer allocate an intermediate string for strings which need to be mutated.

It also opens opportunities for further optimizations:
- no intermediate string allocations for truncation and normalization (directly read the strings from the raw byte buffer)
- string interning (we could intern some strings directly in the decoding layer.

Benchmarks:

```
name              old time/op    new time/op    delta
DecoderMsgpack-4    19.0ms ± 2%    12.8ms ± 1%  -32.30%  (p=0.008 n=5+5)

name              old alloc/op   new alloc/op   delta
DecoderMsgpack-4    9.12MB ± 0%    9.15MB ± 0%   +0.36%  (p=0.008 n=5+5)

name              old allocs/op  new allocs/op  delta
DecoderMsgpack-4      129k ± 0%      129k ± 0%     ~     (all equal)
```

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Write there any instructions and details you may have to test your PR.

### CODE REVIEW

File: pkg/trace/api/api.go [hunk 1]
```diff
@@ -49,10 +48,11 @@ type HTTPReceiver struct {
...
 
-	out     chan *Payload
-	conf    *config.AgentConfig
-	dynConf *sampler.DynamicConfig
-	server  *http.Server
+	out        chan *Payload
+	conf       *config.AgentConfig
+	dynConf    *sampler.DynamicConfig
+	server     *http.Server
+	bufferPool *sync.Pool
```
Comment (gbbr) [inline pkg/trace/api/api.go:55]: I see pools usually being defined on the global scope. Any reason it's a field? Also, I don't think it needs to be a pointer...

File: pkg/trace/api/api.go [hunk 2]
```diff
@@ -331,7 +337,7 @@ func (r *HTTPReceiver) tagStats(v Version, req *http.Request) *info.TagStats {
 	})
 }
 
-func decodeTraces(v Version, req *http.Request) (pb.Traces, error) {
+func decodeTraces(v Version, req *http.Request, bufferPool *sync.Pool) (pb.Traces, error) {
```
Comment (gbbr) [inline pkg/trace/api/api.go:340]: If it's global, you don't need the arg 👍

File: pkg/trace/api/api.go [hunk 3]
```diff
@@ -589,7 +603,15 @@ func decodeRequest(req *http.Request, dest msgp.Decodable) error {
 	default:
 		// do our best
 		if err1 := json.NewDecoder(req.Body).Decode(dest); err1 != nil {
-			if err2 := msgp.Decode(req.Body, dest); err2 != nil {
+			buffer := bufferPool.Get().(*bytes.Buffer)
+			buffer.Reset()
```
Comment (gbbr) [inline pkg/trace/api/api.go:607]: These two can go into a `newBuffer` func

File: pkg/trace/pb/decoder.go [hunk 4]
```diff
@@ -228,3 +228,164 @@ func parseInt32(dc *msgp.Reader) (int32, error) {
 		return 0, msgp.TypeError{Encoded: t, Method: msgp.IntType}
 	}
 }
+
+// parseStringBytes reads the next type in the msgpack payload and
```
Comment (gbbr) [inline pkg/trace/pb/decoder.go:232]: Maybe we should have another file called `decoder_bytes.go`

File: pkg/trace/pb/decoder_test.go [hunk 5]
```diff
@@ -46,24 +46,56 @@ func TestParseFloat64(t *testing.T) {
...
+
+func TestDecodeBytes(t *testing.T) {
 	want := Traces{
 		{{Service: "A", Name: "op"}},
 		{{Service: "B"}},
 		{{Service: "C"}},
 	}
-	var buf bytes.Buffer
-	if err := msgp.Encode(&buf, &want); err != nil {
+	var bts []byte
```
Comment (gbbr) [inline pkg/trace/pb/decoder_test.go:86]: Note that I usually tended to group multiple variable definitions using `var ( ...` as opposed to C-like `var` statements. No biggie, but it would be nice to try and keep it consistent. Arguably I know that there's some old code somewhere which goes against my statement here.

**Review by gbbr — COMMENTED** [review]
Left a few nits. I think we should merge this. Please ping me again once it's out of draft...

File: pkg/trace/api/api.go (hunk 1 above)
Comment (bmermet) [reply pkg/trace/api/api.go:55]: I usually consider it a best practice to avoid variables. But in that case it simplifies slightly the code and I don't see the harm, I'll make the change.

File: pkg/trace/api/api.go (hunk 3 above)
Comment (bmermet) [reply pkg/trace/api/api.go:607]: I see the benefit of factoring the code, but at the same time I like the fact that the `bufferPool.Get()` is explicit and visible in the same function than  `bufferPool.Put()` I think it makes the flow clearer and makes it less likely to copy the code and forget to Put back the buffer in the pool. What do you think?

Comment (gbbr) [reply pkg/trace/api/api.go:607]: The general practice is a global variable for a `sync.Pool` and a `new*` constructor to get items from it and reset. That's what I've always been using and that's the form used in the Go standard library too (you can grep for both `sync.Pool` and `Pool.Get` to see for yourself if you'd like).

On that note, "best practice" can also be subjective, but I think in this case there is enough evidence that the proposed approach is more common.

Comment (gbbr) [reply pkg/trace/api/api.go:607]: Dropping in one of many examples: https://github.com/golang/go/blob/go1.13.10/src/net/http/server.go#L822-L836

File: pkg/trace/api/api.go [hunk 6]
```diff
@@ -43,6 +42,18 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
+var bufferPool = sync.Pool{
+	New: func() interface{} {
+		return new(bytes.Buffer)
+	},
+}
+
+func newBuffer() *bytes.Buffer {
```
Comment (gbbr) [inline pkg/trace/api/api.go:51]: Can we go one step further and have it be `popBuffer` and `pushBuffer`. That'll make it less error prone and might address your concern about forgetting to put it back...

File: pkg/trace/pb/span_gen.go [hunk 7]
```diff
@@ -1,357 +1,290 @@
...
+			var zb0003 uint32
+			zb0003, bts, err = msgp.ReadMapHeaderBytes(bts)
 			if err != nil {
+				err = msgp.WrapError(err, "Metrics")
 				return
 			}
-			if z.Metrics == nil && zhct > 0 {
-				z.Metrics = make(map[string]float64, zhct)
+			if z.Metrics == nil && zb0003 > 0 {
+				z.Metrics = make(map[string]float64, zb0003)
```
Comment (furmmon) [inline pkg/trace/pb/span_gen.go:227]: len(z.Metrics) > 0 below can happen when a field is sent multiple times?

**Review by raphaelgavache — APPROVED** [review]
lgtm

File: pkg/trace/pb/span_gen.go (hunk 7 above)
Comment (bmermet) [reply pkg/trace/pb/span_gen.go:227]: That or if the a message is deserialized in an object that has already a non empty Metrics field.

File: pkg/trace/api/api.go [hunk 8]
```diff
@@ -576,10 +591,17 @@ func (r *HTTPReceiver) Languages() string {
 	return strings.Join(str, "|")
 }
 
-func decodeRequest(req *http.Request, dest msgp.Decodable) error {
+func decodeRequest(req *http.Request, dest *pb.Traces) error {
 	switch mediaType := getMediaType(req); mediaType {
 	case "application/msgpack":
-		return msgp.Decode(req.Body, dest)
+		buffer := getBuffer()
```
Comment (gbbr) [inline pkg/trace/api/api.go:597]:
```suggestion
		buf := getBuffer()
```
🙏

File: pkg/trace/api/api.go [hunk 9]
```diff
@@ -589,7 +611,14 @@ func decodeRequest(req *http.Request, dest msgp.Decodable) error {
 	default:
 		// do our best
 		if err1 := json.NewDecoder(req.Body).Decode(dest); err1 != nil {
-			if err2 := msgp.Decode(req.Body, dest); err2 != nil {
+			buffer := getBuffer()
```
Comment (gbbr) [inline pkg/trace/api/api.go:614]: Same here please.

File: pkg/trace/pb/span_gen_modifs_test.go [hunk 10]
```diff
@@ -0,0 +1,333 @@
+package pb
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/tinylib/msgp/msgp"
+)
+
+// These tests check the custom modifications made on top of the msgpack
```
Comment (gbbr) [inline pkg/trace/pb/span_gen_modifs_test.go:10]: Thanks for doing this!

File: pkg/trace/pb/span_gen_modifs_test.go [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+
+func testStringField(t *testing.T, name string, get func(span *Span) string) {
+	t.Run("Nil", func(t *testing.T) {
+		b := newEmptyMessage()
+		b = msgp.AppendString(b, name)
+		b = msgp.AppendNil(b)
+		s, err := decodeBytes(b)
+		assert.Nil(t, err)
+		assert.Equal(t, "", get(s))
+	})
```
Comment (gbbr) [inline pkg/trace/pb/span_gen_modifs_test.go:33]: Can we have an empty line between tests pls?

File: pkg/trace/test/runner.go [hunk 12]
```diff
@@ -109,18 +108,18 @@ func (s *Runner) Post(traceList pb.Traces) error {
 		return errors.New("post: trace-agent not running")
 	}
 
-	var buf bytes.Buffer
-	if err := msgp.Encode(&buf, traceList); err != nil {
+	bts, err := traceList.MarshalMsg(nil)
```
Comment (gbbr) [inline pkg/trace/test/runner.go:111]: Why this change? I don't think it matters here... Did it break?

File: pkg/trace/pb/decoder_bytes.go [hunk 13]
```diff
@@ -0,0 +1,169 @@
+package pb
```
Comment (gbbr) [inline pkg/trace/pb/decoder_bytes.go:1]: Is this code generated? IIRC you were mentioning that yes. If so, why doesn't it have the right headers? Either way, copyright might be nice.

File: pkg/trace/pb/decoder_bytes.go [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+// parseUint64Bytes parses an uint64 even if the sent value is an int64;
+// this is required because the language used for the encoding library
+// may not have unsigned types. An example is early version of Java
+// (and so JRuby interpreter) that encodes uint64 as int64:
+// http://docs.oracle.com/javase/tutorial/java/nutsandbolts/datatypes.html
+func parseUint64Bytes(bts []byte) (uint64, []byte, error) {
+	// read the generic representation type without decoding
+	t := msgp.NextType(bts)
+
+	var i int64
```
Comment (gbbr) [inline pkg/trace/pb/decoder_bytes.go:116]: Can you please group these?
```
var (
    ...
)
```

File: pkg/trace/pb/decoder_bytes.go (hunk 13 above)
Comment (bmermet) [reply pkg/trace/pb/decoder_bytes.go:1]: As discussed offline this is not generated.

File: pkg/trace/test/runner.go (hunk 12 above)
Comment (bmermet) [reply pkg/trace/test/runner.go:111]: Yes, I removed the io.Writer based serializer so traceLIst no longer implements the `msg.Encodable` interface.

**Review by gbbr — APPROVED** [review]
